prover/src/lib.rs

```diff
@@ -424,109 +431,329 @@ impl<C: SP1ProverComponents> SP1Prover<C> {
-                                shard_proofs,
-                                kinds,
-                                is_complete,
-                            };
+            // Spawn a worker that sends the first layer inputs to a bounded channel.
+            let input_sync = Arc::new(TurnBasedSync::new());
+            let (input_tx, input_rx) = sync_channel::<(usize, usize, SP1CompressMemoryLayouts)>(
+                opts.recursion_opts.checkpoints_channel_capacity,
+            );
+            let input_tx = Arc::new(Mutex::new(input_tx));
+            {
+                let input_tx = Arc::clone(&input_tx);
+                let input_sync = Arc::clone(&input_sync);
+                s.spawn(move || {
+                    for (index, input) in first_layer_inputs.into_iter().enumerate() {
+                        input_sync.wait_for_turn(index);
+                        input_tx.lock().unwrap().send((index, 0, input)).unwrap();
+                        input_sync.advance_turn();
+                    }
+                });
+            }
 
-                            tracing::debug_span!("compress machine proof").in_scope(|| {
-                                self.compress_machine_proof(
-                                    input,
-                                    &self.compress_program,
-                                    &self.compress_pk,
-                                    opts,
-                                )
-                                .map(|p| (p, ReduceProgramType::Reduce))
-                            })
-                        })
-                        .collect::<Vec<_>>()
-                })
-                .collect::<Result<Vec<_>, _>>()?;
+            // Spawn workers who generate the records and traces.
+            let record_and_trace_sync = Arc::new(TurnBasedSync::new());
+            let (record_and_trace_tx, record_and_trace_rx) =
+                sync_channel::<(
+                    usize,
+                    usize,
+                    ExecutionRecord<BabyBear>,
+                    Vec<(String, RowMajorMatrix<BabyBear>)>,
+                    ReduceProgramType,
+                )>(opts.recursion_opts.records_and_traces_channel_capacity);
+            let record_and_trace_tx = Arc::new(Mutex::new(record_and_trace_tx));
+            let record_and_trace_rx = Arc::new(Mutex::new(record_and_trace_rx));
+            let input_rx = Arc::new(Mutex::new(input_rx));
+            for _ in 0..opts.recursion_opts.trace_gen_workers {
+                let record_and_trace_sync = Arc::clone(&record_and_trace_sync);
+                let record_and_trace_tx = Arc::clone(&record_and_trace_tx);
+                let input_rx = Arc::clone(&input_rx);
+                let span = tracing::debug_span!("generate records and traces");
+                s.spawn(move || {
+                    let _span = span.enter();
+                    loop {
+                        let received = { input_rx.lock().unwrap().recv() };
```

Comment:
This might be a very small different but it seems a bit inefficient that we have a Mutex on the Receiver because they would basically compete for the same resource. Another way to do it is to have one loop on the recv() and send tasks to workers, this might reduce context switching but it might also end up being the same.